simulator/func_sim/alu.h

```diff
@@ -10,12 +10,84 @@
 #include "multiplication.h"
 #include "traps/trap.h"
 
+#include <array>
 #include <infra/macro.h>
 #include <tuple>
 
 template<size_t N, typename T>
 T align_up(T value) { return ((value + ((1ull << N) - 1)) >> N) << N; }
 
+struct Mask {
+    const uint32 left_mask, right_mask;
+};
+
+static std::array<Mask, 4> shuffle_consts = {
+    Mask{ 0x00ff0000, 0x0000ff00},
+    Mask{ 0x0f000f00, 0x00f000f0},
+    Mask{ 0x30303030, 0x0c0c0c0c},
+    Mask{ 0x44444444, 0x22222222},
+};
+
+static uint32 shuffle_stage( uint32 src, size_t index)
+{
+    uint32 maskL = shuffle_consts[index].left_mask;
+    uint32 maskR = shuffle_consts[index].right_mask;
+    index = 1 << ( 3 - index);
+    uint32 x = src & ~( maskL | maskR);
+    x |= ( ( src << index) & maskL) | ( ( src >> index) & maskR);
+    return x;
+}
+
+static inline uint32 bit_shuffle( uint32 src1, uint32 src2)
+{
+    uint32 result = src1;
+    uint32 shamt = src2 & 15;
+    for( size_t i = 0; i < 4 ; ++i)
+    {
+        if ( shamt & ( 1 << ( 3 - i)))
+            result = shuffle_stage( result, i);
+    }
+    return result;
+}
+
+static inline uint64 bit_shuffle( uint64 src1, uint64 src2)
+{
+    uint64 result = src1;
+    uint32 shamt = src2 & 31;
+    uint32 left_part = result >> 32;
+    uint32 right_part = result;
```

Comment:
I advise to use `narrow_cast` here to emphasize we are narrowing. Plus, if I recall correctly, we have more sever warnings in Visual Studio.

Reply:
By the way, is it possible to use variables of the same type, just apply bit masks?

Reply:
I later use auxiliary variables in overloaded function `bit_shuffle` and using same type variables will call the wrong candidate.

Reply:
OK